src/arc_utilities/reliable_tf.py

```diff
@@ -0,0 +1,113 @@
+        r.sleep()
+
+
+def send_transform_from_pose_msg(exit_event, tfw, pose, parent, child, is_static):
+    r = Rate(10)
+    while not exit_event.is_set():
+        tfw.send_transform_from_pose_msg(pose, parent, child, is_static, time=rospy.Time.now())
+        r.sleep()
+
+
+class ReliableTF(TF2Wrapper):
+    """
+    abstracts the process of repeatedly sending transforms, to make it look more like sending transforms is guaranteed.
+    this class is a foot gun, it will probably break if you use it in complicated ways.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.managed_tfs = {}
+        self.shared_tfw = TF2Wrapper()
+
+    def __del__(self):
+        self.close()
+
+    def close(self):
+        for parent, child in list(self.managed_tfs.keys()):
+            self.stop_send(parent, child)
+
+    def stop_send(self, parent, child):
+        k = tfkey(parent=parent, child=child)
+        if k in self.managed_tfs:
+            thread, exit_event = self.managed_tfs.pop(k)
+            exit_event.set()
+            thread.join()
+
+    def start_send_transform_matrix(self, transform, parent, child, is_static=False):
+        send_transform_partial = partial(send_transform_matrix, tfw=self.shared_tfw, transform=transform, parent=parent,
+                                         child=child, is_static=is_static)
+
+        def _is_close_func(received_transform):
+            return np.allclose(transform, received_transform)
+
+        self.start_send(parent, child, _is_close_func, send_transform_partial)
+
+    def start_send_transform_from_pose_msg(self, pose: Pose, parent, child, is_static=False):
+        send_transform_partial = partial(send_transform_from_pose_msg, tfw=self.shared_tfw, pose=pose, parent=parent,
+                                         child=child, is_static=is_static)
+
+        def _is_close_func(received_transform):
+            return np.allclose(ros_numpy.numpify(pose), received_transform)
+
+        self.start_send(parent, child, _is_close_func, send_transform_partial)
+
+    def start_send_transform(self, translation, quaternion, parent, child, is_static=False):
+        send_transform_partial = partial(send_transform, tfw=self.shared_tfw, translation=translation,
+                                         quaternion=quaternion, parent=parent,
+                                         child=child, is_static=is_static)
+
+        def _is_close_func(received):
+            received_translation = received[:-1, -1]
+            received_quaternion = tf.transformations.quaternion_from_matrix(received)
+            trans_close = np.allclose(translation, received_translation)
+            rot_close = np.allclose(quaternion, received_quaternion)
+            return trans_close and rot_close
+
+        self.start_send(parent, child, _is_close_func, send_transform_partial)
+
+    def start_send(self, parent, child, is_close_func, send_transform_partial):
```

Comment:
Move next to `stop_send`, or move `stop_send` down so they can be read together?